`Range Function` for rational functions

gxyd opened the pull request:
Implement the `Range Function` for a given `domain` for `Rational Functions`

TODO:
- [x] Rename to `codomain`
  - [x] function name and other mentions in the code
  - [x] docstring
  - [x] filename
- [x] Add `codomain` to `__init__.py` file in `sympy/calculus`.
- [x] Naming Changes.
- [x] Add docstring
- [x] Coverage (98 %) for codomain.py (missing 2) (excluded 2)

Commit:
`Range Function` for rational functions

Commit:
singularity handling corrected

hargup commented:
@gxyd you should move the range function to a new file rather than making the changing in `singularities`.

There are merge conflicts.

hargup commented on sympy/calculus/singularities.py:
```diff
@@ -31,3 +34,130 @@ def singularities(expr, sym):
+    if len(free_symbol) == 1:
+        symbol = free_symbol.pop()
+    elif len(free_symbol) == 0:
+        return FiniteSet(func)
```
Why do we do this?

hargup replied:
(sorry for the nitpick)

gxyd replied:
For `func = 1` which contains `no-symbols` FiniteSet(1) is to be returned if the domain is `non-empty`
`Use case`
```
>>> range_func(1, Interval(1, 2))
FiniteSet(1)
```
@hargup  please let me know if some thing is wrong here, or the point is not clear.

Commit:
doctests added for `range_func` and made `range_func.py`

Commit:
Merge branch 'master' into Range_Function

gxyd commented:
ping @hargup @aktech

gxyd commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,169 @@
+
+    if isinstance(set_value, FiniteSet):
+        set_value = Complement(set_value, FiniteSet(*sing))
+        return FiniteSet(*[limit(func, symbol, i) if not i in sing
```
As i think this line of code can be modified to just:
```
FiniyeSet([*func.subs({symbol: i}) for i in set_value])
```

Commit:
`error` for non-rational functions

Commit:
commiting

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+    sing = closure_handle(set_value, sing)
+
+    def in_intrvl(f, set_val):
+        val1 = (set_val.start, set_val.args[2])
```
`set_val.args[2]` -> `sel_val.left_open`

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+
+    def in_intrvl(f, set_val):
+        val1 = (set_val.start, set_val.args[2])
+        val2 = (set_val.end, set_val.args[3])
```
`set_val.args[3]` -> `sel_val.right_open`

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+from sympy.solvers.solveset import solveset_real
+
+
+def range_func(func, set_value):
```
`codomain`[#1] is a more precise term for the function you are trying to implement here, plus it avoids the confusion with the builtin `range` function.

[#1]: https://en.wikipedia.org/wiki/Codomain

hargup replied:
I would like you to explicitly take the involved variables as a tuple rather than determining them using `func.free_symbols`. As a general principle `explicit is better than implicit` plus in this case it will help us return paratmetric `codomain` say for things like the general quadratic equation.

hargup replied:
`set_value` -> `domain`.

hargup commented:
@gxyd you need to add `codomain` to `__init__.py` file in `sympy/calculus` so that it becomes the part of the public API.

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+
+
+def range_func(func, set_value):
+    """ Finds the range of a real-valued function
```
Thanks for mentioning `real-valued` function.

jksuom replied:
Perhaps something could be said of the argument, too. Real, or more general?

hargup replied:
Agreed, the domain is also a real set, it should also be mentioned.

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+    sing = singularities(func, symbol)
+    sing = closure_handle(set_value, sing)
+
+    def in_intrvl(f, set_val):
```
`in_intrvl` -> `codomain_interval`

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+    def in_intrvl(f, set_val):
+        val1 = (set_val.start, set_val.args[2])
+        val2 = (set_val.end, set_val.args[3])
+        g1 = simplify(diff(f, symbol))
```
`simplify` docstring says:

> Simplification is not a well defined term and the exact strategies
this function tries can change in the future versions of SymPy. If
your algorithm relies on "simplification" (whatever it is), try to
determine what you need exactly  -  is it powsimp()?, radsimp()?,
together()?, logcombine()?, or something else? And use this particular
function directly, because those are well defined and thus your algorithm
will be robust.

`powsimp` or `together` do the simplifications you want to have. If you don't a have specific usecase and you are using `simplify` as a "just in case" solution it is better to not to use it.

hargup replied:
`g1` is a very uninformative name better use `df` and for `g2` below you can use `d2f`.

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+        mini = set()
+        ans1 = limit(f, symbol, val1[0])
+        ans2 = limit(f, symbol, val2[0], '-')
+        singl = solveset_real(1/f, symbol)
```
Why don't you use `sing` calculated above?

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+
+        maxi = set()
+        mini = set()
+        ans1 = limit(f, symbol, val1[0])
```
defining `val1` and `val2` doesn't help, please directly use `set_val.start`/`set_val.end`, `set_val.left_open`/set_val.right_open`.

hargup replied:
`ans1` -> `start_val`

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+        maxi = set()
+        mini = set()
+        ans1 = limit(f, symbol, val1[0])
+        ans2 = limit(f, symbol, val2[0], '-')
```
https://github.com/sympy/sympy/pull/9719/files#r35408348

hargup replied:
`ans2` -> `end_val`

hargup commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+
+        for i in singl:
+            if i in sing and not i in set_val.boundary:
+                return Union(range_func(f, Interval(val1[0], i, val1[1], True)),
```
https://github.com/sympy/sympy/pull/9719/files#r35408348

hargup commented:
@gxyd Here is some advise for naming functions/variables in code. 

* If the reviewer has to scroll up to see the definition of the variable then it is probably a terrible name for that variable.
* Prefer full words or even pharses over acronyms and hard to guess shortforms.
* Use a code completion tool if typing long names is a pain for you. You can use [YouCompleteMe](https://valloric.github.io/YouCompleteMe/) if you are using vim.

aktech commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+    ======
+
+    NotImplementedError
+          The algorithms for to find the solution of the given equation are
```
`for to` --> `to`

aktech commented on sympy/calculus/range_func.py:
```diff
@@ -0,0 +1,170 @@
+        raise ValueError(" %s is not a valid sympy expression" % (func))
+
+    if not isinstance(set_value, Set):
+        raise ValueError('A Symbol must be given, not type %s: %s' % (type(set_value), set_value))
```
`A Symbol` --> `A Set`

Commit:
function name changed from `range_func` to `codomain`

gxyd commented:
ping @hargup  @aktech

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,166 @@
+
+    func = sympify(func)
+    if not isinstance(domain, Set):
+        raise ValueError('A Set must be given, not type %s: %s' % (type(domain), domain))
```
`not type %s` -> `not %s`

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,166 @@
+    Examples
+    ========
+
+    >>> from sympy.calculus.codomain import codomain
```
Since you have added `codomain` to `__init__.py`, just doing `from sympy.calculus import codomain` should do.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,166 @@
+            else:
+                mini = set([(start_val, set_val.left_open and set_val.right_open)])
+
+        unk = set()
```
Please expand `unk`.

gxyd replied:
OK. I will change it to `d2f_zero_val` though earlier it meant `unknown for d2f`.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,166 @@
+        der_zero = solveset_real(df1, symb)
+        der_zero_in_dom = closure_handle(set_val, der_zero)
+
+        maxi = set()
```
`maxi` -> `maxima` or `local_maxima`
`mini` -> `minima` or `local_minima`

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,166 @@
+            else:
+                unk.add(f.subs({symb: i}))
+
+        ma = (-oo, True)
```
`ma` -> `maximum`
`mi` -> `minimum`

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,166 @@
+        if start_val is S.Infinity or end_val is S.Infinity:
+            maxi = set([(oo, True)])
+        elif start_val is S.NegativeInfinity or end_val is S.NegativeInfinity:
+            mini = set([(-oo, True)])
```
A new line after this line.

hargup commented:
You need not handle singularities in `codomain_interval` if you remove the singularities from the domain before the control reaches it.

gxyd commented:
But even if i remove the singularities. The limit still needs to be taken for them isn't it ? So removing `singularities`  from `domain` makes `no difference` other than for `FiniteSet` for which `limit` is `not used`.
Like
```
>>> codomain(x/(x - 4), Interval(1, 4), x)
(-oo, S(-1)/3]
>>> codomain(x/(x -4), Interval(1, 4, False, True), x)
(-oo, S(-1)/3]            # same result after removing `4` from Interval 
```
Here even if change `Interval(1, 4)` to `Interval(1, 4, False, True)` i will have to check whether `4` will be approached from the `-ve` side or `+ve` direction of `4`, in this case `1 < 4`  =>  `-ve` limit should be taken.
@hargup Am i wrong somewhere ?

Commit:
`spacing` added `unk` removed

hargup commented:
That's right but if you remove the singularities from the domain, then they will only appear on the boundaries so you can do away [this](https://github.com/sympy/sympy/pull/9719/files#diff-1299a2752b6796eed94f8cb6510761f2R99) block of code.

gxyd commented:
Ok.  Is there anything more to be addressed here @hargup ?

hargup commented:
> Ok. Is there anything more to be addressed here @hargup ?

Everything else looks good, thanks for your work.

Commit:
`singularities` removed initially

gxyd commented:
>  if you remove the singularities from the domain

That also helped in reducing the `time of tests` from `4.3 sec` to `3.95 sec` .

aktech commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,159 @@
+        raise NotImplementedError("more than one variables %s not handeled" % (syms))
+
+    if not func.is_rational_function(symbol):
+        raise NotImplementedError("Algorithms finding range for non-rational functions"
```
Put a space after `functions` otherwise the concatenated string would look like:
`Algorithms finding range for non-rational` **`functionsare`** `not yet implemented`

aktech commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,159 @@
+    if len(syms) == 1:
+        symbol = syms[0]
+    else:
+        raise NotImplementedError("more than one variables %s not handeled" % (syms))
```
handeled --> handled

aktech replied:
`% (syms)` ---> `% (syms,)`

Otherwise it would though a `TypeError`.

gxyd replied:
Ya i noticed this one. Thanks.

Commit:
`typos` fixed in `codomain`

aktech commented on sympy/calculus/codomain.py:
```diff
@@ -0,0 +1,159 @@
+        raise ValueError('A Set must be given, not %s: %s' % (type(domain), domain))
+
+    if len(syms) == 0:
+        raise ValueError("A Symbol or a tuple of symbols must be given: not %s" % (syms,))
```
Simply: `raise ValueError("A Symbol or a tuple of symbols must be given)` , Since `syms` is empty, no need to format it's type into the string.

This change `(syms) --  >(syms,)` was suggested for only line 54, while formatting a tuple into a string.

gxyd replied:
> Simply: raise ValueError("A Symbol or a tuple of symbols must be given) , Since syms is empty, no need to format it's type into the string.

Ok, i think that be better.

Commit:
error message changed

aktech commented:
![calculus](https://cloud.githubusercontent.com/assets/5647941/8894100/38b38d76-33c8-11e5-8e87-fc18e54a6a48.png)

[Coverage Report](http://iamit.in/sympy/coverage-report/calculus/sympy_calculus_codomain_py.html) Looks good, Can you add some tests for few untested lines (4), like [this](http://iamit.in/sympy/coverage-report/calculus/sympy_calculus_codomain_py.html#n127)?

gxyd commented:
Its `cool` i did not knew this `feature of coverage report`. I thought this is just for `GSOC` participants mentioning what they `have to do`.
Thanks @aktech

aktech commented:
In future if you ever want get the report, just run:

```
bin/coverage_report.py sympy/calculus
```

Commit:
`solveset_real` used instead of `singularities`

I have now removed the some conditions which are only required for the
non-rational functions. Using `solveset_real` is better compared to the
`singularities`

gxyd commented:
@aktech  I have now have the coverage report [here](https://medium.com/@gauravdhingra/coverage-report-files-for-calculus-bdae89928599)

aktech commented:
> In singularities expressions like
>
> `expr = (x**2 - 1)/(x**3 - 1)`
>are first simplified to (x + 1)/(x**2 + x + 1) which is wrong

That seems like a bug.

>  I have added the test case here which was previously giving wrong result.

Tests looks good. Thanks!

aktech commented:
> I have now have the coverage report here

Looks Great! +1

hargup commented:
Thanks! +1. Please also update the [release notes](https://github.com/sympy/sympy/wiki/Release-Notes-for-0.7.7).

gxyd commented:
Thanks @hargup  @aktech . Ok i will update the release notes.

asmeurer commented:
Why was this renamed from range to codomain? I think range is a more accurate name, as it is generally defined as the image of the domain (what this calculates). For instance, you can have f: R -> R defined by f(x) = sin(x). The codomain of f is R but the range is [-1, 1].  See https://en.wikipedia.org/wiki/Codomain.

I think we should rename this.

gxyd commented:
@asmeurer  please have a look at [this](https://github.com/sympy/sympy/pull/9719#discussion_r35406149) comment by @hargup . Personally i felt `range_func` to be the best name. I disagree with `range` as the name. As for now, we can have its name as `image` which may also be similar `imageset` which we have in SymPy.

asmeurer commented:
I disagree that codomain is the correct term for this. 

What actually is the difference between this function and imageset? It seems that they are very similar.

hargup commented:
I agree with @asmeurer here, sorry for the confusion I created. By definition this code is same as `imageset` for rational function and it appears that the new functionality provided by this code is already in `imageset`. Now I'm wondering why didn't I point that out during the code review.

gxyd commented:
@asmeurer @hargup  i have opened issue #9967 for removing this function.

aktech commented:
@gxyd  Sorry for your wasted efforts.
